test/server/test-run-test.js

```diff
@@ -0,0 +1,248 @@
+            fixture: 'fixture',
+            fn:      noop
+        };
+
+        var testRun          = new TestRun(testMock);
+        var click1           = commands.createClickCommand();
+        var click2           = commands.createClickCommand();
+        var commandsToClient = [];
+        var testRunDone      = false;
+
+        testRun.once('done', function () {
+            testRunDone = true;
+        });
+
+        var doneCommandResult = {
+            status: 'done'
+        };
+
+        var actionError         = 'action-error';
+        var failedCommandResult = {
+            status: 'failed',
+            err:    actionError
+        };
+
+        function sendClientRequest (commandResult) {
+            return testRun[CLIENT_MESSAGES.ready](commandResult)
+                .then(function (command) {
+                    commandsToClient.push(command);
+                });
+        }
+
+        sendClientRequest()
+            .then(function () {
+                // NOTE: simulate page reload before pending step is
+                // executed that leads to repeated ready message.
+                return sendClientRequest();
+            })
+            .then(function () {
+                return sendClientRequest(doneCommandResult);
+            })
+            .then(function () {
+                return sendClientRequest(failedCommandResult);
+            });
+
+        return testRun.executeCommand(click1)
+            .then(function () {
+                return testRun.executeCommand(click2);
+            })
+            .then(function () {
+                throw new Error('command should be rejected');
+            })
+            .catch(function (err) {
+                expect(err).equal(actionError);
+
+                testRun._done();
+                return nextTick();
+            })
+            .then(function () {
+                expect(testRunDone).to.be.true;
+                expect(commandsToClient.length).eql(4);
+                expect(commandsToClient[0]).eql(click1);
+                expect(commandsToClient[1]).eql(click1);
+                expect(commandsToClient[2]).eql(click2);
+                expect(commandsToClient[3].type).eql(commands.TYPE.testDone);
+            });
+    });
+
+    it('Should call test.fn() and complete the test after it is resolved', function () {
+        var testMock = {
+            fixture: 'fixture',
+            fn:      Promise.resolve
+        };
+
+        var testRun     = new TestRun(testMock);
+        var testRunDone = false;
+
+        testRun.once('done', function () {
+            testRunDone = true;
+        });
+
+        return testRun[CLIENT_MESSAGES.ready]()
+            .then(function (command) {
+                expect(command.type).eql(commands.TYPE.testDone);
+                expect(testRunDone).to.be.true;
+            });
+    });
+
+    it('Should call test.fn() and complete the test with an error after it is rejected', function () {
```

Comment:
Should call test.fn() and complete the test with an error after the function is rejected